llvm/lib/Transforms/Scalar/CorrelatedValuePropagation.cpp

```diff
@@ -1289,6 +1299,14 @@ static bool runImpl(Function &F, LazyValueInfo *LVI, DominatorTree *DT,
     FnChanged |= BBChanged;
   }
 
+  // Infer range attribute on return value.
+  if (RetRange && !RetRange->isFullSet()) {
+    Attribute RangeAttr = F.getRetAttribute(Attribute::Range);
+    if (RangeAttr.isValid())
+      RetRange = RetRange->intersectWith(RangeAttr.getRange());
+    if (!RetRange->isEmptySet() && !RetRange->isSingleElement())
```

Comment:
I've excluded the single element case to make this less spammy. I didn't want to have range attributes on every single function returning a constant (in which case IPSCCP would already propagate the cosntant).

Reply:
Maybe leave that as a comment?